lib/opencensus/tags/tag_map.rb

```diff
@@ -1,71 +1,116 @@
 # frozen_string_literal: true
 
+# Copyright 2019 OpenCensus Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
 
 require "forwardable"
+require "opencensus/tags/tag"
 
 module OpenCensus
   module Tags
     # # TagMap
     #
-    # Collection of tag key and value.
+    # Collection of tags that can be used to label anything that is
+    # associated with a stats or trace operations.
+    #
     # @example
     #
-    #  tag_map = OpenCensus::Tags::OpenCensus.new
+    #  tag_map = OpenCensus::Tags::TagMap.new
+    #
+    #  # Add tag
+    #  tag_map << OpenCensus::Tags::Tag.new "key1", "val1"
     #
-    #  # Add or update
-    #  tag_map["key1"] = "value1"
-    #  tag_map["key2"] = "value2"
+    #  # Add tag with TagTTL
+    #  tag_map << OpenCensus::Tags::Tag.new "key2", "val2", ttl: 0
     #
-    #  # Get value
-    #  tag_map["key1"] # value1
+    #  # Update tag
+    #  tag_map << OpenCensus::Tags::Tag.new "key3", "val3"
+    #  tag_map << OpenCensus::Tags::Tag.new "key3", "updatedval3"
+    #
+    #  # Get tag by key
+    #  p tag_map["key1"]
+    #  # <OpenCensus::Tags::Tag:0x007ffc138 @key="key1", @value="val1", @ttl=-1>
     #
     #  # Delete
     #  tag_map.delete "key1"
     #
     #  # Iterate
-    #  tag_map.each do |key, value|
-    #    p key
-    #    p value
+    #  tag_map.each do |tag|
+    #    p tag.key
+    #    p tag.value
     #  end
     #
     #  # Length
     #  tag_map.length # 1
     #
-    # @example Create tag map from hash
+    # @example Create tag map with list of tags.
     #
-    #   tag_map = OpenCensus::Tags::OpenCensus.new({ "key1" => "value1"})
+    #   tag_map = OpenCensus::Tags::OpenCensus.new([
+    #     OpenCensus::Tags::Tag.new "key1", "val1",
+    #     OpenCensus::Tags::Tag.new "key2", "val2"
+    #   ])
     #
     class TagMap
       extend Forwardable
 
-      # The maximum length for a tag key and tag value
-      MAX_LENGTH = 255
-
-      # Create a tag map. It is a map of tags from key to value.
-      # @param [Hash{String=>String}] tags Tags hash with string key and value.
+      # Create a tag map.
       #
-      def initialize tags = {}
-        @tags = {}
+      # @param [Array<Tags::Tag>] tags Tags list with string key and value and
+      #   metadata.
+      #
+      def initialize tags = []
+        @tags = tags.each_with_object({}) { |tag, r| r[tag.key] = tag }
```

Comment:
Perhaps there should also be a way to initialize a TagMap quickly by passing a `Hash<String,String>`. So that, for example, when I initialize a Measurement, I can pass in `Measurement.new tags: {"key1" => "value1"}` rather than having to go through the ceremony of constructing a `TagMap` first.

Reply:
As per the tag metadata spec, Tag have additional properties like TTL for tag prorogation, so to manage key, value and metadata fields.
My suggestion that tags can be created using both way

```ruby
Measurement.new tags: {"key1" => "value1"}
# OR
tag_map = TagMap.new
tag_map << Tag.create_non_propagative "key", "value", ttl: 
Measurement.new tags: tag_map
```

@dazuma Let me know your thoughts?

Reply:
Yes, I agree, we should have both options.

Reply:
Added create tag map with both Hash and Tags array